localtrack.go

```diff
@@ -526,16 +527,22 @@ func (s *LocalTrack) rtcpWorker(rtcpReader interceptor.RTCPReader) {
 }
 
 func (s *LocalTrack) writeWorker(provider SampleProvider, onComplete func()) {
+	s.lock.Lock()
 	if s.cancelWrite != nil {
 		s.cancelWrite()
+		// wait for previous write to finish to prevent multi-threaded provider reading
+		<-s.writeClosed
```

Comment:
Nice fix for locking `s.cancelWrite`. though I'm a bit concerned about holding the lock here while waiting for the previous write to complete.  What do you think about the following instead?

```
s.cancelWrite()
s.cancelWrite = nil
s.lock.Unlock()

<-s.writeClosed
```

Reply:
Ah yes, the locking! It's a tricky one for me. I don't see any ways to make this truly thread-safe, without linking the waiting and locking together.

----
This would be your suggestion in context, if I understand correctly?
```
func (s *LocalTrack) writeWorker(provider SampleProvider, onComplete func()) {
	s.lock.Lock()
	if s.cancelWrite != nil {
		s.cancelWrite()
		s.cancelWrite = nil
		s.lock.Unlock()
		
		<-s.writeClosed		
		s.lock.Lock()
	}
```

Setting `cancelWrite = nil` would cause a third call to `writeWorker` skip waiting for the first call while the second call is waiting for the first one.

---
Another idea:
```
func (s *LocalTrack) writeWorker(provider SampleProvider, onComplete func()) {
	s.lock.Lock()
	if s.cancelWrite != nil {
		s.cancelWrite()
		s.lock.Unlock()
		<-s.writeClosed
		s.lock.Lock()
	}

```
But then, two calls to `writeWorker` could get stuck waiting for the original worker to finish. Then, when the original writer finishes, one of the waiting writers would get the lock and initialize. But then the next waiting writer would not kill the second writer waiting worker, and would start running parallel to the new worker.

---

The only option to not occupy the `s.lock` during restarts of the writeWorker, would be to introduce a new lock specifically for this function.

Am I missing something? I'm a bit out of idea's on this one :sweat:

Reply:
I think I've solved the puzzle: https://github.com/livekit/server-sdk-go/pull/428/commits/b0cfd345a7e8497f26a43deea34fbb8533d26869